python/test_infra/ttir_builder.py

```diff
@@ -0,0 +1,224 @@
+            typ = input.type
+        elif isinstance(input, OpView):
+            typ = input.operation.result.type
+        elif isinstance(input, Operation):
+            typ = input.result.type
+        else:
+            raise TypeError(f"Invalid input {type(input)}")
+
+        assert isinstance(typ, RankedTensorType), "Only ranked tensors are supported"
+
+        return typ
+
+    def get_shape(self, input: Operand):
+        """Retrieves shape of operand which is expected to be a shaped type."""
+        return self.__get_type(input).shape
+
+    # ----- Utility factories -----
+
+    def ranked_tensor(
+        self,
+        shape: Shape,
+        data_type: Optional[Type] = None,
+        encoding: Optional[Attribute] = None,
+    ) -> RankedTensorType:
+        """Convenience wrapper constructing `RankedTensorType`."""
+        dtype = data_type if data_type is not None else self.__default_dtype
+        with self.ctx, self.loc:
+            return RankedTensorType.get(shape, dtype, encoding)
+
+    def empty(
+        self,
+        shape: Shape,
+        data_type: Optional[Type] = None,
+    ) -> OpView:
+        """Convenience wrapper constructing `tensor.EmptyOp`."""
+        dtype = data_type if data_type is not None else self.__default_dtype
+        with self.ctx, self.loc:
+            return tensor.EmptyOp(shape, dtype)
+
+    # ----- TTIR op factories -----
+
+    def add(self, in0: Operand, in1: Operand) -> OpView:
+        """Convenience wrapper constructing `ttir.AddOp`."""
+        assert self.get_shape(in0) == self.get_shape(
+            in1
+        ), "Elementwise `ttir.add` op expects inputs of same shape."
+
+        with self.ctx, self.loc:
+            output = self.empty(self.get_shape(in0))
+
+            return ttir.AddOp(
+                [self.__get_type(output)],
+                [in0, in1],
+                [output],
+                self.__get_operand_constraint_attr(3),
+            )
+
+    def multiply(self, in0: Operand, in1: Operand) -> OpView:
+        """Convenience wrapper constructing `ttir.MultiplyOp`."""
+        assert self.get_shape(in0) == self.get_shape(
+            in1
+        ), "Elementwise `ttir.multiply` op expects inputs of same shape."
+
+        with self.ctx, self.loc:
+            output = self.empty(self.get_shape(in0))
+
+            return ttir.MultiplyOp(
+                [self.__get_type(output)],
+                [in0, in1],
+                [output],
+                self.__get_operand_constraint_attr(3),
+            )
+
+    def exp(self, in0: Operand) -> OpView:
+        """Convenience wrapper constructing `ttir.ExpOp`."""
+        with self.ctx, self.loc:
+            output = self.empty(self.get_shape(in0))
+
+            return ttir.ExpOp(
+                [self.__get_type(output)],
+                [in0],
+                [output],
+                self.__get_operand_constraint_attr(3),
+            )
+
+
+def wrap(*inputs_shapes: Tuple[Shape]):
```

Comment:
`wrap` might be a bit too generic of a name.  Maybe it should be called `compile` or `build` or `func`? lmk what you think.

Reply:
I settled with verbose `compile_as_mlir_module`.